### ISSUE TITLE
Collect tutorials' performance data on NV platform

### ISSUE THREAD

**Issue by tdeng5** [intel/intel-xpu-backend-for-triton#544]
Please help collect Softmax, GEMM, flash attention's performance data based on A100 platform for our reference.

Tutorial list: 
https://triton-lang.org/main/getting-started/tutorials/02-fused-softmax.html
https://triton-lang.org/main/getting-started/tutorials/03-matrix-multiplication.html
https://triton-lang.org/main/getting-started/tutorials/06-fused-attention.html
https://triton-lang.org/main/getting-started/tutorials/11-grouped-gemm.html

**Comment by whitneywhtsang** [comment]
```
$ python 02-fused-softmax.py
softmax-performance:
          N      Triton  Torch (native)  Torch (jit)
0     256.0  546.133347      546.133347   199.804881
1     384.0  558.545450      558.545450   193.511801
2     512.0  585.142849      585.142849   197.397593
3     640.0  605.693176      620.606056   195.047621
4     768.0  630.153868      630.153868   196.607991
..      ...         ...             ...          ...
93  12160.0  675.555532      351.181430   226.747520
94  12288.0  675.628857      351.713770   223.927109
95  12416.0  675.521202      351.603543   223.334452
96  12544.0  675.771063      351.188110   223.660666
97  12672.0  675.839968      351.085730   223.664650

[98 rows x 4 columns]
```
```
$ python 03-matrix-multiplication.py
triton_output=tensor([[-30.4844,   7.5391, -11.2500,  ..., -16.4531,  -5.3398,   3.8965],
        [ 52.5938, -11.8125, -29.0312,  ..., -32.6250, -23.9531, -27.7812],
        [ 22.6562, -21.3438,   1.1152,  ...,  -2.9609, -78.3750,  34.2812],
        ...,
        [-13.5234,  27.5469,  -8.5234,  ...,  -6.9727,   9.0781,  18.9219],
        [  1.8652,   9.3516,   6.7500,  ..., -16.6250, -21.3125, -19.4531],
        [-17.5938, -28.5938,   0.5913,  ...,  22.0000,  -3.4648,  43.1250]],
       device='cuda:0', dtype=torch.float16)
torch_output=tensor([[-30.4844,   7.5391, -11.2500,  ..., -16.4531,  -5.3398,   3.8965],
        [ 52.5938, -11.8125, -29.0312,  ..., -32.6250, -23.9531, -27.7812],
        [ 22.6562, -21.3438,   1.1152,  ...,  -2.9609, -78.3750,  34.2812],
        ...,
        [-13.5234,  27.5469,  -8.5234,  ...,  -6.9727,   9.0781,  18.9219],
        [  1.8652,   9.3516,   6.7500,  ..., -16.6250, -21.3125, -19.4531],
        [-17.5938, -28.5938,   0.5913,  ...,  22.0000,  -3.4648,  43.1250]],
       device='cuda:0', dtype=torch.float16)
✅ Triton and Torch match
matmul-performance:
         M       N       K     cuBLAS     Triton
0    256.0   256.0   256.0   5.461333   5.461333
1    384.0   384.0   384.0  13.824000  13.824000
2    512.0   512.0   512.0  25.890766  28.387843
3    640.0   640.0   640.0  36.408890  46.545454
4    768.0   768.0   768.0  42.130286  44.306030
5    896.0   896.0   896.0  63.860363  63.860363
6   1024.0  1024.0  1024.0  72.315584  83.886082
7   1152.0  1152.0  1152.0  66.355199  74.475050
8   1280.0  1280.0  1280.0  65.015874  75.983769
9   1408.0  1408.0  1408.0  79.011245  92.354069
10  1536.0  1536.0  1536.0  76.933564  77.778988
11  1664.0  1664.0  1664.0  87.368079  86.985407
12  1792.0  1792.0  1792.0  76.458667  76.982353
13  1920.0  1920.0  1920.0  87.493673  84.809814
14  2048.0  2048.0  2048.0  83.468735  95.325090
15  2176.0  2176.0  2176.0  80.173899  85.632545
16  2304.0  2304.0  2304.0  74.417047  80.975836
17  2432.0  2432.0  2432.0  77.823997  88.070416
18  2560.0  2560.0  2560.0  77.649287  83.805625
19  2688.0  2688.0  2688.0  83.552988  90.316801
20  2816.0  2816.0  2816.0  81.218262  87.909722
21  2944.0  2944.0  2944.0  77.747321  84.040530
22  3072.0  3072.0  3072.0  83.514905  89.030036
23  3200.0  3200.0  3200.0  79.900126  86.720870
24  3328.0  3328.0  3328.0  78.895923  84.003845
25  3456.0  3456.0  3456.0  79.118323  85.404201
26  3584.0  3584.0  3584.0  75.432376  83.720102
27  3712.0  3712.0  3712.0  77.024641  85.154902
28  3840.0  3840.0  3840.0  76.060522  81.497422
29  3968.0  3968.0  3968.0  76.504033  84.621313
30  4096.0  4096.0  4096.0  89.003801  89.151600
```
```
$ python 06-fused-attention.py
fused-attention-batch4-head48-d64-fwd-causal=True:
     N_CTX     Triton
0   1024.0  48.257350
1   2048.0  57.611712
2   4096.0  65.147791
3   8192.0  68.096259
4  16384.0  69.831567
fused-attention-batch4-head48-d64-fwd-causal=False:
     N_CTX     Triton
0   1024.0  67.405833
1   2048.0  68.684449
2   4096.0  70.600655
3   8192.0  70.688961
4  16384.0  71.642094
fused-attention-batch4-head48-d64-bwd-causal=True:
     N_CTX     Triton
0   1024.0  33.543651
1   2048.0  42.991448
2   4096.0  50.055212
3   8192.0  54.399408
4  16384.0  56.924383
```
```
$ python 11-grouped-gemm.py
group-gemm-performance:
        N    cuBLAS    Triton
0   128.0  0.014336  0.012288
1   256.0  0.017408  0.016384
2   512.0  0.036864  0.025600
3  1024.0  0.113744  0.106496
```

**Comment by whitneywhtsang** [comment]
@tdeng5 Please confirm if the collected data is good.

**Comment by tdeng5** [comment]
The data looks good. For softmax, could we get the full data list? Need a couple of lines change for the script, Yudong will comment.

**Comment by ESI-SYD** [comment]
Just for reference: patch run in 100 times and store data in csv
```
diff --git a/python/tutorials/02-fused-softmax.py b/python/tutorials/02-fused-softmax.py
index b083be79..be0360ac 100644
--- a/python/tutorials/02-fused-softmax.py
+++ b/python/tutorials/02-fused-softmax.py
@@ -20,7 +20,9 @@ In doing so, you will learn about:
 #
 # Custom GPU kernels for elementwise additions are educationally valuable but won't get you very far in practice.
 # Let us consider instead the case of a simple (numerically stabilized) softmax operation:
-
+import csv
+import os
+import statistics
 import torch
 import intel_extension_for_pytorch  # type: ignore # noqa: F401
 
@@ -160,13 +162,9 @@ assert torch.allclose(y_triton, y_torch), (y_triton, y_torch)
         line_arg='provider',  # argument name whose value corresponds to a different line in the plot
         line_vals=[
             'triton',
-            'torch-native',
-            'torch-jit',
         ],  # possible values for `line_arg``
         line_names=[
             "Triton",
-            "Torch (native)",
-            "Torch (jit)",
         ],  # label name for the lines
         styles=[('blue', '-'), ('green', '-'), ('green', '--')],  # line styles
         ylabel="GB/s",  # label name for the y-axis
@@ -176,17 +174,37 @@ assert torch.allclose(y_triton, y_torch), (y_triton, y_torch)
 def benchmark(M, N, provider):
     x = torch.randn(M, N, device='xpu', dtype=torch.float32)
     quantiles = [0.5, 0.2, 0.8]
-    if provider == 'torch-native':
-        ms, min_ms, max_ms = triton.testing.do_bench(lambda: torch.softmax(x, axis=-1), quantiles=quantiles)
-    if provider == 'triton':
-        ms, min_ms, max_ms = triton.testing.do_bench(lambda: softmax(x), quantiles=quantiles)
-    if provider == 'torch-jit':
-        ms, min_ms, max_ms = triton.testing.do_bench(lambda: naive_softmax(x), quantiles=quantiles)
-    gbps = lambda ms: 2 * x.nelement() * x.element_size() * 1e-9 / (ms * 1e-3)
-    return gbps(ms), gbps(max_ms), gbps(min_ms)
-
-
-benchmark.run(show_plots=True, print_data=True)
+    gbps_values = []
+    for i in range(100):
+        if provider == 'torch-native':
+            ms, min_ms, max_ms = triton.testing.do_bench(lambda: torch.softmax(x, axis=-1), quantiles=quantiles)
+        if provider == 'triton':
+            ms, min_ms, max_ms = triton.testing.do_bench(lambda: softmax(x), quantiles=quantiles)
+        if provider == 'torch-jit':
+            ms, min_ms, max_ms = triton.testing.do_bench(lambda: naive_softmax(x), quantiles=quantiles)
+        gbps = lambda ms: 2 * x.nelement() * x.element_size() * 1e-9 / (ms * 1e-3)
+        print(f"============Result: round {i} for N {N}=======")
+        print(f"============gbps(ms): {gbps(ms)}, gbps(max_ms):{gbps(max_ms)}, gbps(min_ms): {gbps(min_ms)}=======")
+
+        gbps_values.append(gbps(ms))
+    
+    min_gbps = min(gbps_values)
+    max_gbps = max(gbps_values)
+    median_gbps = statistics.median(gbps_values)
+    
+    # Store data in CSV file
+    filename = 'benchmark_results.csv'
+    file_exists = os.path.isfile(filename)
+    with open(filename, mode='a') as file:
+        writer = csv.writer(file)
+        if not file_exists:
+            writer.writerow(['N', 'min_gbps', 'max_gbps', 'median_gbps'])
+        writer.writerow([N, min_gbps, max_gbps, median_gbps])
+    
+    return min_gbps, max_gbps, median_gbps
+
+
+benchmark.run(show_plots=False, print_data=True)
 
 # %%
 # In the above plot, we can see that:


```

**Comment by tdeng5** [comment]
please keep "Torch (native)", "Torch (jit)" data when you are collecting, looks like yudong's patch didn't include them.

**Comment by whitneywhtsang** [comment]
@tdeng5 Just noticed that the measurements are done on a machine with 2 NVIDIA RTX A5000, not A100. I don't have access to a A100 machine.

**Comment by tdeng5** [comment]
A5000's data is fine, we just need to collect some data for reference. I updated the issue's title.

**Comment by whitneywhtsang** [comment]
As discussed, the collected data is good enough.
